File: tmdb-core/src/main/java/dev/reuss/tmdb/domain/trending/model/TrendingAllResult.java
package dev.reuss.tmdb.domain.trending.model;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import dev.reuss.tmdb.common.TmdbModel;
import dev.reuss.tmdb.util.TmdbCollections;
import dev.reuss.tmdb.value.media.MediaType;

import java.util.List;

/**
 * Result item returned by TMDB trending all.
 *
 * <p>This model supports movie, TV and person results. Fields that do not
 * apply to the concrete media type may be {@code null} or empty.</p>
 *
 * @param adult            whether the result is marked as adult
 * @param backdropPath     backdrop image path
 * @param id               TMDB resource id
 * @param title            movie title
 * @param originalTitle    original movie title
 * @param name             TV show or person name
 * @param originalName     original TV show or person name
 * @param originalLanguage original language code
 * @param overview         overview
 * @param posterPath       poster image path
 * @param profilePath      person profile image path
 * @param mediaType        media type, for example {@code movie}, {@code tv} or {@code person}
 * @param genreIds         genre ids
 * @param popularity       popularity
 * @param releaseDate      movie release date
 * @param firstAirDate     TV first air date
 * @param video            whether the movie has video content
 * @param voteAverage      vote average
 * @param voteCount        vote count
 * @param originCountry    TV origin countries
 */
@JsonIgnoreProperties(ignoreUnknown = true)
public record TrendingAllResult(
        boolean adult,

        @JsonProperty("backdrop_path")
        String backdropPath,

        int id,

        String title,

        @JsonProperty("original_title")
        String originalTitle,

        String name,

        @JsonProperty("original_name")
        String originalName,

        @JsonProperty("original_language")
        String originalLanguage,

        String overview,

        @JsonProperty("poster_path")
        String posterPath,

        @JsonProperty("profile_path")
        String profilePath,

        @JsonProperty("media_type")
        MediaType mediaType,

        @JsonProperty("genre_ids")
        List<Integer> genreIds,

        double popularity,

        @JsonProperty("release_date")
        String releaseDate,

        @JsonProperty("first_air_date")
        String firstAirDate,

        boolean video,

        @JsonProperty("vote_average")
        double voteAverage,

        @JsonProperty("vote_count")
        int voteCount,

        @JsonProperty("origin_country")
        List<String> originCountry
) implements TmdbModel {

    public TrendingAllResult {
        genreIds = TmdbCollections.list(genreIds);
        originCountry = TmdbCollections.list(originCountry);
    }
}

File: tmdb-core/src/main/kotlin/dev/reuss/tmdb/domain/trending/model/TrendingAllResult.kt
package dev.reuss.tmdb.domain.trending.model

import com.fasterxml.jackson.annotation.JsonProperty
import dev.reuss.tmdb.common.TmdbModel
import dev.reuss.tmdb.value.media.MediaType

/**
 * Result item returned by TMDB trending all.
 *
 * This model supports movie, TV and person results. Fields that do not
 * apply to the concrete media type may be `null` or empty.
 *
 * @property adult whether the result is marked as adult
 * @property backdropPath backdrop image path
 * @property id TMDB resource id
 * @property title movie title
 * @property originalTitle original movie title
 * @property name TV show or person name
 * @property originalName original TV show or person name
 * @property originalLanguage original language code
 * @property overview overview
 * @property posterPath poster image path
 * @property profilePath person profile image path
 * @property mediaType media type, for example `movie`, `tv` or `person`
 * @property genreIds genre ids
 * @property popularity popularity
 * @property releaseDate movie release date
 * @property firstAirDate TV first air date
 * @property video whether the movie has video content
 * @property voteAverage vote average
 * @property voteCount vote count
 * @property originCountry TV origin countries
 */
@JvmRecord
data class TrendingAllResult(
    @all:JsonProperty("adult")
    val adult: Boolean,
    @all:JsonProperty("backdrop_path")
    val backdropPath: String?,
    @all:JsonProperty("id")
    val id: Int,
    @all:JsonProperty("title")
    val title: String?,
    @all:JsonProperty("original_title")
    val originalTitle: String?,
    @all:JsonProperty("name")
    val name: String?,
    @all:JsonProperty("original_name")
    val originalName: String?,
    @all:JsonProperty("original_language")
    val originalLanguage: String?,
    @all:JsonProperty("overview")
    val overview: String?,
    @all:JsonProperty("poster_path")
    val posterPath: String?,
    @all:JsonProperty("profile_path")
    val profilePath: String?,
    @all:JsonProperty("media_type")
    val mediaType: MediaType?,
    @all:JsonProperty("genre_ids")
    val genreIds: List<Int> = emptyList(),
    @all:JsonProperty("popularity")
    val popularity: Double,
    @all:JsonProperty("release_date")
    val releaseDate: String?,
    @all:JsonProperty("first_air_date")
    val firstAirDate: String?,
    @all:JsonProperty("video")
    val video: Boolean,
    @all:JsonProperty("vote_average")
    val voteAverage: Double,
    @all:JsonProperty("vote_count")
    val voteCount: Int,
    @all:JsonProperty("origin_country")
    val originCountry: List<String> = emptyList(),
) : TmdbModel
